app-experimental.tex

```diff
@@ -2693,6 +2693,125 @@ \section{Distributed Capabilities For Peripherals And Accelerators} % <<<
 
 % >>>
 % >>>
+\section{Thread Identification} % <<<
+
+\subsection{Motivation}
+
+Compartmentalisation models often rely on trusted code performing the switch from one compartment to another.
+This trusted code needs a mechanism to retrieve a data capability, e.g., for a stack.
+When using sentries (sealed entry capabilities) to call into trusted code, the same piece of trusted code can be called from multiple threads and each thread must have a distinct trusted data capability.
+Thus, we have the need for reliable thread identification.
+This can be solved by calling into the entity that provides threading support.
+In the conventional case, this would mean a call to kernel space at every time trusted code is entered.
+Thus, we need to have a way of both fast and reliably determining the current thread ID (TID).
+
+The TID is not the same as Thread-Local Storage (TLS).
+When jumping from one compartment to another, the TLS should change per compartment, but the TID should remain the same over all calls.
+The commonly used RISC-V calling convention defines \texttt{x4} as the thread pointer register.
+We cannot use \texttt{x4} as the TID register because it is not a reliable source.
+Any compartment can manipulate it.
+
+\subsection{Design}
+
+We currently envision multiple layers to this design.
+
+\begin{enumerate}
+\item The most simple design is to have one thread in the system called the Supervisor Thread ID (STID).
```

Comment:
“In the system” seems a very broad scope. Do we mean per core / hardware context / HART / something like that?

Reply:
I should change that to hardware thread. That also is what the RISC-V terminology seems to be.

Reply:
Userspace (generally) isn’t dealing with hardware threads though, it’s dealing with software threads. You don’t want your software thread ID changing when you get migrated to another hart, nor do you want a different software thread to start using yours if it later runs on the same hart.